src/executor/insert.rs

```diff
@@ -137,29 +138,49 @@ where
 {
     type PrimaryKey<A> = <<A as ActiveModelTrait>::Entity as EntityTrait>::PrimaryKey;
     type ValueTypeOf<A> = <PrimaryKey<A> as PrimaryKeyTrait>::ValueType;
-    let last_insert_id_opt = match db.support_returning() {
-        true => {
-            let cols = PrimaryKey::<A>::iter()
-                .map(|col| col.to_string())
-                .collect::<Vec<_>>();
-            let rows = db.query_all(statement).await?;
-            let res = rows.last().ok_or_else(|| {
-                DbErr::RecordNotInserted("None of the records are being inserted".to_owned())
-            })?;
-            res.try_get_many("", cols.as_ref()).ok()
-        }
-        false => {
-            let last_insert_id = db.execute(statement).await?.last_insert_id();
-            ValueTypeOf::<A>::try_from_u64(last_insert_id).ok()
+
+    enum QueryOrExecResult {
+        Query(QueryResult),
+        Exec(ExecResult),
+    }
+
+    let insert_result = if db.support_returning() {
+        let mut rows = db.query_all(statement).await?;
+        if rows.is_empty() {
+            return Err(DbErr::RecordNotInserted(
+                "None of the records are being inserted".to_owned(),
+            ));
         }
+        QueryOrExecResult::Query(rows.remove(rows.len() - 1))
+    } else {
+        QueryOrExecResult::Exec(db.execute(statement).await?)
     };
+
     let last_insert_id = match primary_key {
-        Some(value_tuple) => FromValueTuple::from_value_tuple(value_tuple),
-        None => match last_insert_id_opt {
-            Some(last_insert_id) => last_insert_id,
-            None => return Err(DbErr::UnpackInsertId),
-        },
-    };
+        Some(value_tuple) => Ok(FromValueTuple::from_value_tuple(value_tuple)),
```

Comment:
It seems to be we just have to add a `db.execute(statement).await?` here rather than introduce an entirely new type.

Reply:
No. It has to execute whether Some or None.

Reply:
So, if we add the execute statement in the `Some` branch, we can inline the top of the function in the `None` branch, further reducing the amount of work done in case we provide the primary key. Each branch would have its own singular await point.

What's the rationale for reducing the number of await points?